File: 559. Maximum Depth of N-ary Tree/MaxDepth.py
'''
https://leetcode.com/problems/maximum-depth-of-n-ary-tree/
'''


# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class Solution:
    # O(n) time,
    # O(n) space,
    # Approach: DFS,
    def maxDepth(self, root: 'Node') -> int:
        if not root:
            return 0
        
        stack = [(root, 1)]
        max_depth = 1
        
        while stack:
            nod, depth = stack.pop()
            max_depth = max(max_depth, depth)
        
            for child in nod.children:
                stack.append((child, depth+1))
        
        return max_depth

File: 1614. Maximum Nesting Depth of the Parentheses/max_depth.rs

use std::cmp::max;


impl Solution {
    
    // O(n) time,
    // O(1) space,
    // Approach: stack, counting
    pub fn max_depth(s: String) -> i32 {
        let mut pairs = 0;
        let mut ans = 0;
        
        for ch in s.chars() {
            match ch {
                '(' => pairs += 1,
                ')' => pairs -=1,
                _ => continue
            }
            ans = max(ans, pairs);
        }
        
        ans
    }
}